### PR TITLE
Single pass gas estimation.

### PR DESCRIPTION
**Pull request by StefanIliev545**
### Why this change is needed

Gas estimation currently is doing a binary search for the most optimal gas. This maxes it out and makes it run at maximum returning the gas used. Inefficient for complex contracts that have non deterministic logic based on provided gas, but those should potentially be handled through a separate rpc. 

### What changes were made as part of this PR

Removed the binary search doEstimateGas in favour of estimateGasSinglePass

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: go/config/enclave_cli_flags.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ var EnclaveFlags = map[string]*flag.TenFlag{
 	MessageBusAddressFlag:         flag.NewStringFlag(MessageBusAddressFlag, "", "The address of the L1 message bus contract owned by the management contract."),
 	MaxBatchSizeFlag:              flag.NewUint64Flag(MaxBatchSizeFlag, 1024*55, "The maximum size a batch is allowed to reach uncompressed"),
 	MaxRollupSizeFlag:             flag.NewUint64Flag(MaxRollupSizeFlag, 1024*128, "The maximum size a rollup is allowed to reach"),
-	L2BaseFeeFlag:                 flag.NewUint64Flag(L2BaseFeeFlag, params.InitialBaseFee, ""),
+	L2BaseFeeFlag:                 flag.NewUint64Flag(L2BaseFeeFlag, params.InitialBaseFee/10, ""),
```
Comment (tudor-malene) [inline go/config/enclave_cli_flags.go:56]: ah, so this is a config.
Then it's better to set the value we want in the config.
Maybe coordinate with @BedrockSquirrel

File: go/enclave/evm/evm_facade.go [hunk 2]
```diff
@@ -362,30 +362,33 @@ func ExecuteObsCall(
 		noBaseFee = false
 	}
 
+	ethHeader, err := gethEncodingService.CreateEthHeaderForBatch(ctx, header)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshot := s.Snapshot()
+	defer s.RevertToSnapshot(snapshot) // Always revert after simulation
```
Comment (tudor-malene) [inline go/enclave/evm/evm_facade.go:371]: nice one. We should have probably had this a long time

File: go/enclave/rpc/EstimateGas.go [hunk 3]
```diff
@@ -94,7 +93,7 @@ func EstimateGasExecute(builder *CallBuilder[CallParamsWithBlock, hexutil.Uint64
 	// TODO: Change to fixed time period quotes, rather than this.
 	publishingGas = publishingGas.Mul(publishingGas, gethcommon.Big2)
 
-	executionGasEstimate, gasPrice, err := rpc.doEstimateGas(builder.ctx, txArgs, blockNumber, rpc.config.GasLocalExecutionCapFlag)
+	executionGasEstimate, gasPrice, err := rpc.estimateGasSinglePass(builder.ctx, txArgs, blockNumber, rpc.config.GasLocalExecutionCapFlag)
```
Comment (tudor-malene) [inline go/enclave/rpc/EstimateGas.go:96]: needs a high-level comment of the approach maybe with a simple example.
I find it quite hard to follow what's going on. This stuff is a bit confusing

**Review by tudor-malene — COMMENTED** [review]
I think it looks good.
But I'd like to read a high level comment with the approach

File: go/enclave/rpc/EstimateGas.go (hunk 3 above)
Comment (StefanIliev545) [reply go/enclave/rpc/EstimateGas.go:96]: It needs a bit of refactoring tbh, but I guess we can go with a comment for now.

Comment (StefanIliev545) [reply go/enclave/rpc/EstimateGas.go:102]: Added some comments in a couple of places.

**Review by tudor-malene — COMMENTED** [review]
looks good, but there is an issue since the e2e tests are failing

**Review by tudor-malene — APPROVED** [review]
lgtm
